torch/csrc/distributed/rpc/python_functions.cpp

```diff
@@ -101,7 +127,6 @@ py::object toPyObjInternal(RpcCommandBase& rpc, MessageType messageType) {
       auto& resp = static_cast<PythonResp&>(rpc);
       auto& pythonRpcHandler = PythonRpcHandler::getInstance();
       py::object ret = pythonRpcHandler.deserialize(resp.serializedPyObj());
-      pythonRpcHandler.handleException(ret);
```

Comment:
what's wrong with handleException?

Reply:
I moved it to `unwrap_func` that I added for PythonFutureWrapper.

The exception must be raise in the client thread calling the pybind
"wait" API, so that it can be correctly shown to user. 

Calling it here is wrong, here it's raised in RPC server thread, where the exception would be
swallowed in the ThreadPool task, and also no pybind handling code can help shown the Python exception.

Reply:
1, calling `toPyObj(..)` could raise a Python exception to C++ land.

2, C++ land by default does not handle Python exceptions, unless the C++ code is wrapped into `py::cpp_function`, which is auto-generated for us when we do `py::module::def()`.
https://github.com/pytorch/pytorch/blob/a2557970f3eefe4c4ec3ff30b7d1c33d8b7bc246/torch/csrc/distributed/rpc/init.cpp#L255


This all happens on client side. But even for client side, it uses it's server thread to receive response and deserialize the bytes into a Message object.
https://github.com/pytorch/pytorch/blob/340ccf56fb2b4f0739082f528b3fc8d12d568c20/torch/csrc/distributed/rpc/process_group_agent.cpp#L499-L500

Before this change, server doesn't further deserialzie the Message object into a Python object. It was done in user Python thread.

After this change, specifically the wrapFutureMessageInIvalueFuture, we are doing deserializa Message to py::object in server thread. This align our goal to parallel embedding bag lookup, deserialization ideally should be put into server thread pool, so deserialization for multiple objects can happen in parallel, rather than doing it in user Python thread where only one object can be in deserialization at a time. 

While application user is in a Python main thread and calls the wait(), so the work left to it is to raise the exception if necessary.